src/main/scala/firrtl/Namespace.scala

```diff
@@ -63,22 +63,20 @@ object Namespace {
   def apply(m: DefModule): Namespace = {
     val namespace = new Namespace
 
-    def buildNamespaceStmt(s: Statement): Statement =
-      s map buildNamespaceStmt match {
-        case dec: IsDeclaration =>
-          namespace.namespace += dec.name
-          dec
-        case x => x
-      }
-    def buildNamespacePort(p: Port): Port = p match {
-      case dec: IsDeclaration =>
-        namespace.namespace += dec.name
-        dec
-      case x => x
+    def buildNamespaceStmt(s: Statement): Seq[String] = s match {
+      case s: IsDeclaration => Seq(s.name)
+      case s: Conditionally => buildNamespaceStmt(s.conseq) ++ buildNamespaceStmt(s.alt)
+      case s: Block => s.stmts flatMap buildNamespaceStmt
+      case _ => Nil
     }
-    m.ports map buildNamespacePort
+    def buildNamespacePort(p: Port): Seq[String] = p match {
+      case dec: IsDeclaration => Seq(dec.name)
+      case x => Nil
+    }
+    namespace.namespace ++= (m.ports flatMap buildNamespacePort)
```

Comment:
This is more idiomatic then it was but we can actually do better. What you really mean here is Option not Seq since you only ever have 1, you could instead change it to 
```scala
    def buildNamespacePort(p: Port): Option[String] = p match {
      case dec: IsDeclaration => Some(dec.name)
      case _ => None
    }
```
It turns out that this pattern of matching on a type and returning Some for it and None as default is really common so there is a builtin function for it (typical Scala) called `collect`.
Thus this entire thing can be more concisely be written as:
```scala
  namespace.namespace ++= (m.ports collect { case dec: IsDeclaration => dec.name })
```
Collect is kind of like a cooler filter.